File: leetcode/maximumDepthOfBinaryTree.js
/**
Given a binary tree, find its maximum depth.

The maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.

Note: A leaf is a node with no children.

Example:

Given binary tree [3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7
return its depth = 3.
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  if (!root) {
    return 0;
  }
  if (!root.left) {
    return maxDepth(root.right) + 1;
  }
  if (!root.right) {
    return maxDepth(root.left) + 1;
  }
  return maximum(maxDepth(root.right), maxDepth(root.left)) + 1;
};

const maximum = (first, second) => {
  return first > second ? first : second;
};

File: maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    depth := 0
    if root == nil {
        return depth
    }
    
    children := []*TreeNode{root.Left, root.Right}
    for _, child := range children {
        if child != nil {
            depth = max(depth, maxDepth(child))
        }
    }
    return depth + 1
}

func max(i, j int) int {
    if i > j {
        return i
    }
    return j
}